lib/mayaUsd/ufe/UsdUndoMaterialCommands.cpp

```diff
@@ -375,43 +418,162 @@ void UsdUndoAssignNewMaterialCommand::redo()
         auto addMaterialCmd
             = std::dynamic_pointer_cast<MAYAUSD_NS::ufe::UsdUndoAddNewPrimCommand>(*cmdsIt++);
         auto addShaderCmd = std::dynamic_pointer_cast<UsdUndoCreateFromNodeDefCommand>(*cmdsIt);
-        connectShaderToMaterial(addShaderCmd->insertedChild(), addMaterialCmd->newPrim());
+        connectShaderToMaterial(addShaderCmd->insertedChild(), addMaterialCmd->newPrim(), _nodeId);
     }
 }
 
-void UsdUndoAssignNewMaterialCommand::connectShaderToMaterial(
-    Ufe::SceneItem::Ptr shaderItem,
-    UsdPrim             materialPrim)
+void UsdUndoAssignNewMaterialCommand::markAsFailed()
 {
-    auto                 shaderUsdItem = std::dynamic_pointer_cast<UsdSceneItem>(shaderItem);
-    auto                 shaderPrim = UsdShadeShader(shaderUsdItem->prim());
-    UsdShadeOutput       materialOutput;
-    PXR_NS::SdrRegistry& registry = PXR_NS::SdrRegistry::GetInstance();
+    _cmds->undo();
+    _cmds.reset();
+}
+
+UsdUndoAddNewMaterialCommand::UsdUndoAddNewMaterialCommand(
+    const UsdSceneItem::Ptr& parentItem,
+    const std::string&       nodeId)
+    : Ufe::InsertChildCommand()
+    , _parentPath((parentItem && parentItem->prim().IsActive()) ? parentItem->path() : Ufe::Path())
+    , _nodeId(nodeId)
+    , _cmds(std::make_shared<Ufe::CompositeUndoableCommand>())
+{
+}
+
+UsdUndoAddNewMaterialCommand::~UsdUndoAddNewMaterialCommand() { }
+
+UsdUndoAddNewMaterialCommand::Ptr
+UsdUndoAddNewMaterialCommand::create(const UsdSceneItem::Ptr& parentItem, const std::string& nodeId)
+{
+    // Changing the hierarchy of invalid items is not allowed.
+    if (!parentItem || !parentItem->prim().IsActive())
+        return nullptr;
+
+    return std::make_shared<UsdUndoAddNewMaterialCommand>(parentItem, nodeId);
+}
+
+Ufe::SceneItem::Ptr UsdUndoAddNewMaterialCommand::insertedChild() const
+{
+    if (_cmds) {
+        auto cmdsIt = _cmds->cmdsList().begin();
+        std::advance(cmdsIt, _createMaterialCmdIdx + 1);
+        auto addShaderCmd = std::dynamic_pointer_cast<UsdUndoCreateFromNodeDefCommand>(*cmdsIt);
+        return addShaderCmd->insertedChild();
```

Comment:
And again, it assumes the 2nd child can successfully be cast and the pointer will be valid.

Reply:
Looks like you are doing the review of PR #2641 which indeed completely breaks any assumption about command ordering in UsdUndoAssignNewMaterialCommand. You need to select prims in two separate stages to trigger the bad outcome, which is a broken `redo()` that fails to connect items in the second stage, and the final selection that will only contain the first created shader and not the ones in the other stages.

This will have to be fixed (along with quite a few other commands and workflows that would be baffled by a global selection containing items from multiple stages, especially if there are some well placed name clashes).